components/esp_modem/examples/modem_tcp_client/main/sock_dce.cpp

```diff
@@ -0,0 +1,405 @@
+    if (s == 0) {
+        ESP_LOGD(TAG, "perform select timeout...");
+        return true;
+    } else if (s < 0) {
+        ESP_LOGE(TAG,  "select error %d", errno);
+        close_sock();
+        return false;
+    }
+    if (FD_ISSET(sock, &fdset)) {
+        ESP_LOGD(TAG,  "socket read: data available");
+        if (!signal.wait(IDLE, 1000)) {
+            ESP_LOGE(TAG,  "Failed to get idle");
+            close_sock();
+            return false;
+        }
+        if (state != status::IDLE) {
+            ESP_LOGE(TAG,  "Unexpected state %d", state);
+            close_sock();
+            return false;
+        }
+        state = status::SENDING;
+        int len = ::recv(sock, &buffer[0], size, 0);
+        if (len < 0) {
+            ESP_LOGE(TAG,  "read error %d", errno);
+            close_sock();
+            return false;
+        } else if (len == 0) {
+            ESP_LOGE(TAG,  "EOF %d", errno);
+            close_sock();
+            return false;
+        }
+        ESP_LOG_BUFFER_HEXDUMP(TAG, &buffer[0], len, ESP_LOG_VERBOSE);
+        data_to_send = len;
+        send_cmd("AT+CIPSEND=0," + std::to_string(len) + "\r");
+    }
+    if (FD_ISSET(data_ready_fd, &fdset)) {
+        uint64_t data;
+        read(data_ready_fd, &data, sizeof(data));
+        ESP_LOGD(TAG, "select read: modem data available %x", data);
+        if (!signal.wait(IDLE, 1000)) {
+            ESP_LOGE(TAG, "Failed to get idle");
+            close_sock();
+            return false;
+        }
+        if (state != status::IDLE) {
+            ESP_LOGE(TAG, "Unexpected state %d", state);
+            close_sock();
+            return false;
+        }
+        state = status::RECEIVING;
+        send_cmd("AT+CIPRXGET=2,0," + std::to_string(size) + "\r");
+    }
+    return true;
+}
+
+void DCE::forwarding(uint8_t *data, size_t len)
+{
+    ESP_LOG_BUFFER_HEXDUMP(TAG, data, len, ESP_LOG_DEBUG);
+    if (state == status::SENDING) {
+        if (memchr(data, '>', len) == NULL) {
+            ESP_LOGE(TAG, "Missed >");
+            state = status::SENDING_FAILED;
+            signal.set(IDLE);
+            return;
+        }
+        auto written = dte->write(&buffer[0], data_to_send);
+        if (written != data_to_send) {
+            ESP_LOGE(TAG, "written %d (%d)...", written, len);
+            state = status::SENDING_FAILED;
+            signal.set(IDLE);
+            return;
+        }
+        data_to_send = 0;
+        uint8_t ctrl_z = '\x1A';
+        dte->write(&ctrl_z, 1);
+        state = status::SENDING_1;
+        return;
+    } else if (state == status::RECEIVING || state == status::RECEIVING_1 ) {
+        const int MIN_MESSAGE = 6;
+        size_t actual_len = 0;
+        char *recv_data = (char *)data;
+        if (data_to_recv == 0) {
+            const std::string_view head = "+CIPRXGET: 2,0,";
+            auto head_pos = (char *)std::search(data, data + len, head.begin(), head.end());
```

Comment:
Have you considered making a `string_view` from data as an alternative to the usage of `memchr` and pointer manipulation?

Reply:
I have, but since the replay contains not only string delimiters, but also and mainly data (and that could be very long data), I'd prefer using `memchr` and treating this as raw binary, rather than `string_view`.
As you can see, some portions of those data are parsed using `string_view`, but that doesn't universally work for the entire data we receive.